lib/Frontend/ArgsToFrontendInputsConverter.cpp

```diff
@@ -158,6 +177,15 @@ ArgsToFrontendInputsConverter::createInputFilesConsumingPrimaries(
   bool hasAnyPrimaryFiles = !primaryFiles.empty();
 
   FrontendInputsAndOutputs result;
+  if (LinesToExecute) {
+    bool isPrimary = primaryFiles.count("-e") > 0;
+    result.addInput(InputFile("-e", isPrimary, LinesToExecute.get()));
+    if (isPrimary)
+      primaryFiles.erase("-e");
+
+    Files.remove("-e");
+    result.LinesToExecute = LinesToExecute;
```

Comment:
I'm confused as to what this is doing, but I think it's okay to treat `-e` mode as whole-module, i.e. "no primary files".

Reply:
This block of code suggests to me that there is probably a better approach. What do -e arguments have to do with primary files? Something is not quite right to my eyes here. I *think* this code is using a filename of "-e" as a placeholder for code kept in `LinesToExecute`. But I'm thinking a far better approach would be to generalize `InputFile` to `InputFileOrLineToExecute`, since that's the model you're going after. And put the code to execute right in the `InputFileOrLineToExecute` structure, along with the corresponding modules needed. How does this sound?

Keeping redundant information in both `LinesToExecute` and the InputFile names (-e) feels like it's asking for trouble downstream, sorry.

Reply:
They probably don't have much to do with primary files—I'm probably just confused about when there is a primary file and when there isn't.

`FrontendInputsAndOutputs:: LinesToExecute` is entirely for lifetime management—it is never accessed for anything else. If I comment it out and change the `LinesToExecute` in `ArgsToFrontendInputsConverter` to use a `unique_ptr`, the only difference is that I have to leak the buffer instead of releasing it cleanly.

We ultimately want the lines to execute to be turned into a single `llvm::MemoryBuffer` managed by `SourceManager`. I found `ArgsToFrontendInputsConverter` to be a natural place to concatenate the strings and form a buffer, and `InputFile::Buffer` to be a natural way to get that buffer into the compiler, but we could of course do it in some other way.

Reply:
It's *either* `LinesToExecute` that bother me. The problem is that `LinesToExecute` may have no content, yet there may be an InputFile with a "-e" name. Or there may be content in `LinesToExecute` and no `InputFile` with a "-e". Or many. Yes, one can add assertions, and checks, and diagnostics, but it seems there must be a better way.
For example, if any given frontend job should have either input files or only one -e, then there could be a class triad with a common superclass, one subclass to hold the input files, and a sibling subclass holding the lines to execute. Let's find a way to make code more directly represent what we want it to do.

Reply:
I realized today that the underlying problem here is that if `InputFile` has a `Buffer`, it doesn't take ownership of it. So I think the most general solution is actually to make `InputFile::Buffer` be a `std::shared_ptr` instead of a plain pointer. Is that an acceptable solution? The vast majority of `InputFile`s do not include a `Buffer`; the exceptions are:

* Code generated by the Extract Function refactoring.
* Code generated by applying fix-its in the migrator.
* The main input of `sil-opt`, `sil-func-extractor`, `sil-llvm-gen`, and `sil-nm`.
* `-e` processing.

In all other cases, `Buffer` would be `nullptr`, which I believe would save us from a lot of the overhead that would usually keep us from using a `shared_ptr`.

Does that make sense to you? Or should we try to find some way to transfer ownership of the `Buffer` to a long-lived object more quickly?